File: NeuronMCP/src/middleware/chain.ts
/**
 * Middleware chain execution
 */

import { MCPRequest, MCPResponse, Middleware } from "./types.js";

export class MiddlewareChain {
	constructor(private middlewares: Middleware[]) {
		// Sort by order
		this.middlewares.sort((a, b) => (a.order || 0) - (b.order || 0));
	}

	/**
	 * Execute middleware chain
	 */
	async execute(request: MCPRequest, finalHandler: () => Promise<MCPResponse>): Promise<MCPResponse> {
		let index = 0;
		const enabledMiddlewares = this.middlewares.filter((m) => m.enabled !== false);

		const next = async (): Promise<MCPResponse> => {
			if (index >= enabledMiddlewares.length) {
				return finalHandler();
			}

			const middleware = enabledMiddlewares[index++];
			return middleware.handler(request, next);
		};

		return next();
	}
}

File: NeuronMCP/internal/middleware/chain.go
package middleware

import (
	"context"
	"sort"
)

// Chain executes middleware in order
type Chain struct {
	middlewares []Middleware
}

// NewChain creates a new middleware chain
func NewChain(middlewares []Middleware) *Chain {
	// Sort by order
	sorted := make([]Middleware, len(middlewares))
	copy(sorted, middlewares)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Order() < sorted[j].Order()
	})

	return &Chain{middlewares: sorted}
}

// Execute executes the middleware chain
func (c *Chain) Execute(ctx context.Context, req *MCPRequest, finalHandler Handler) (*MCPResponse, error) {
	// Filter enabled middlewares
	enabled := make([]Middleware, 0)
	for _, mw := range c.middlewares {
		if mw.Enabled() {
			enabled = append(enabled, mw)
		}
	}

	// Build chain
	index := 0
	var next Handler
	next = func(ctx context.Context) (*MCPResponse, error) {
		if index >= len(enabled) {
			return finalHandler(ctx)
		}
		mw := enabled[index]
		index++
		return mw.Execute(ctx, req, next)
	}

	return next(ctx)
}

